python/dgl/nn/pytorch/conv/gatedgraphconv.py

```diff
@@ -75,13 +80,17 @@ def forward(self, graph, feat, etypes):
         graph = graph.local_var()
         zero_pad = feat.new_zeros((feat.shape[0], self._out_feats - feat.shape[1]))
         feat = th.cat([feat, zero_pad], -1)
-        # NOTE(zihao): there is still room to optimize, we may do kernel fusion
-        # for such operations in the future.
-        graph.edata['w'] = self.edge_embed(etypes).view(-1, self._out_feats, self._out_feats)
+
         for _ in range(self._n_steps):
-            graph.ndata['h'] = feat.unsqueeze(-1) # (N, D, 1)
-            graph.update_all(fn.u_mul_e('h', 'w', 'm'),
-                             fn.sum('m', 'a'))
-            a = graph.ndata.pop('a').sum(dim=1) # (N, D)
+            graph.ndata['h'] = feat
+            for i in range(self._n_etypes):
+                eids = (etypes == i).nonzero().view(-1)
+                if len(eids) > 0:
+                    graph.apply_edges(
+                        lambda edges: {'W_e*h': self.linears[i](edges.src['h'])},
+                        eids
+                    )
+            graph.update_all(fn.copy_e('W_e*h', 'm'), fn.sum('m', 'a'))
```

Comment:
Why is this a better implementation?

Reply:
In this way, we do not have to move store `W` on edges, thus reducing the GPU memory consumption, see #852 .

Reply:
I see. From the math, it is quite similar to RGCN. Maybe we could try use heterograph APIs later.